File: 11_动态规划/lis最长上升子序列问题/LIS模板.py
"""
贪心 + 二分查找
LIS[i]表示长度为 i+1 的子序列尾部元素的值
每次遍历到一个新元素,用二分查找法找到第一个大于等于它的元素,然后更新LIS
"""
# LIS模板

from typing import List, Tuple
from bisect import bisect_left, bisect_right


def LIS(nums: List[int], isStrict=True) -> int:
    """求LIS长度"""
    n = len(nums)
    if n <= 1:
        return n

    lis = []  # lis[i] 表示长度为 i 的上升子序列的最小末尾值
    for i in range(n):
        pos = bisect_left(lis, nums[i]) if isStrict else bisect_right(lis, nums[i])
        if pos == len(lis):
            lis.append(nums[i])
        else:
            lis[pos] = nums[i]

    return len(lis)


def getLIS(nums: List[int], isStrict=True) -> Tuple[List[int], List[int]]:
    """求LIS 返回(LIS,LIS的组成下标)"""
    n = len(nums)

    lis = []  # lis[i] 表示长度为 i 的上升子序列的最小末尾值
    dpIndex = [0] * n  # 每个元素对应的LIS长度
    for i in range(n):
        pos = bisect_left(lis, nums[i]) if isStrict else bisect_right(lis, nums[i])
        if pos == len(lis):
            lis.append(nums[i])
        else:
            lis[pos] = nums[i]
        dpIndex[i] = pos

    res, resIndex = [], []
    j = len(lis) - 1
    for i in range(n - 1, -1, -1):
        if dpIndex[i] == j:
            res.append(nums[i])
            resIndex.append(i)
            j -= 1
    return res[::-1], resIndex[::-1]


def caldp(nums: List[int], isStrict=True) -> List[int]:
    """求以每个位置为结尾的LIS长度(包括自身)"""
    if not nums:
        return []

    n = len(nums)
    res = [0] * n
    lis = []
    for i in range(n):
        pos = bisect_left(lis, nums[i]) if isStrict else bisect_right(lis, nums[i])
        if pos == len(lis):
            lis.append(nums[i])
            res[i] = len(lis)
        else:
            lis[pos] = nums[i]
            res[i] = pos + 1
    return res


if __name__ == "__main__":
    assert LIS([10, 9, 2, 5, 3, 7, 101, 18]) == 4
    print(caldp([10, 9, 2, 5, 3, 7, 101, 18]))
    assert getLIS([10, 9, 2, 5, 3, 7, 101, 18]) == ([2, 3, 7, 18], [2, 4, 5, 7])

File: 11_动态规划/lis最长上升子序列问题/LIS模板.ts
import { FastSet } from '../../24_高级数据结构/珂朵莉树/FastSet'
import { bisectLeft, bisectRight } from '../../9_排序和搜索/二分/bisect'

function LIS(nums: ArrayLike<number>, isStrict = true): number {
  const n = nums.length
  if (n <= 1) return n

  const lis: number[] = []
  const bisect = isStrict ? bisectLeft : bisectRight
  for (let i = 0; i < n; i++) {
    const pos = bisect(lis, nums[i])
    if (pos === lis.length) {
      lis.push(nums[i])
    } else {
      lis[pos] = nums[i]
    }
  }

  return lis.length
}

/**
 * 求LIS 返回(LIS,LIS的组成下标).
 */
function getLIS(nums: ArrayLike<number>, isStrict = true): [lis: number[], lisIndex: number[]] {
  const n = nums.length

  const lis: number[] = [] // lis[i] 表示长度为 i 的上升子序列的最小末尾值
  const dpIndex = new Uint32Array(n) // 每个元素对应的LIS长度
  const bisect = isStrict ? bisectLeft : bisectRight
  for (let i = 0; i < n; i++) {
    const pos = bisect(lis, nums[i])
    if (pos === lis.length) {
      lis.push(nums[i])
    } else {
      lis[pos] = nums[i]
    }
    dpIndex[i] = pos
  }

  const res: number[] = []
  const resIndex: number[] = []
  let j = lis.length - 1
  for (let i = n - 1; i >= 0; i--) {
    if (dpIndex[i] === j) {
      res.push(nums[i])
      resIndex.push(i)
      j -= 1
    }
  }

  return [res.reverse(), resIndex.reverse()]
}

/**
 * O(nloglogmax)求`严格递增`的LIS,要求所有元素范围在`[0, max]`内.
 * @param nums 数组.
 * @param max 数组的最大值.不超过1e9.
 */
function LIS2Strict(nums: ArrayLike<number>, max: number): number {
  max += 5
  const alive = new FastSet(max)
  let res = 0

  for (let i = 0; i < nums.length; i++) {
    const cur = nums[i]
    const next = alive.next(cur)
    if (next < max) {
      alive.erase(next)
      alive.insert(cur)
    } else {
      alive.insert(cur)
      res++
    }
  }
  return res
}

export { LIS, LIS2Strict, getLIS }

if (require.main === module) {
  // https://leetcode.cn/problems/longest-increasing-subsequence/
  // eslint-disable-next-line no-inner-declarations
  function lengthOfLIS(nums: number[]): number {
    for (let i = 0; i < nums.length; i++) nums[i] += 1e4
    return LIS2Strict(nums, 2e4)
    return LIS(nums)
  }

  const arr = [1, 2, 2, 3, 3, 3, 1]
  console.log(LIS2Strict(arr, 10))
  console.log(LIS2Strict(arr, 10))
  console.log(getLIS(arr))
}
